### PR TITLE
GEOPY-194

### PR DESCRIPTION
**Pull request by domfournier**
**GEOPY-194 - Remove dependency on the Root group** 
 - Load a geoh5 without a Root group.
- Add unit test

### CODE REVIEW

File: geoh5py/objects/octree.py [hunk 1]
```diff
@@ -194,18 +194,28 @@ def octree_cells(self) -> Optional[np.ndarray]:
 
     @octree_cells.setter
     def octree_cells(self, value):
+
         if value is not None:
-            value = np.vstack(value)
+            if isinstance(value, np.recarray):
```
Comment (sebhmg) [inline geoh5py/objects/octree.py:199]: **question (out-of-scope):** these changes do not seem related to the issue about Root group. Revert from this Pull Request?

File: poetry.lock [hunk 2]
```diff
@@ -8,17 +8,16 @@ python-versions = "*"
 
 [[package]]
```
Comment (sebhmg) [inline poetry.lock:9]: **question (out-of-scope):** revert this file changes from the PR?

File: geoh5py/workspace/workspace.py [hunk 3]
```diff
@@ -103,19 +103,29 @@ def __init__(self, h5file: str = "Analyst.geoh5", **kwargs):
...
 
-            # Get the Root attributes
-            attributes, type_attributes, _ = H5Reader.fetch_attributes(
-                self.h5file, uuid.uuid4(), "Root"
-            )
-            self._root = self.create_entity(
-                RootGroup, save_on_creation=False, **{**attributes, **type_attributes}
-            )
+            try:
+                self._root = self.load_entity(uuid.uuid4(), "root")
```
Comment (sebhmg) [inline geoh5py/workspace/workspace.py:107]: **suggestion:**  have a dedicated function (`fetch_or_create_root()` ?) that encapsulates the whole logic about getting or creating root

File: geoh5py/workspace/workspace.py [hunk 4]
```diff
@@ -742,6 +738,38 @@ def list_objects_name(self) -> Dict[uuid.UUID, str]:
                 objects_name[key] = entity.name
         return objects_name
 
+    def load_entity(self, uid: uuid.UUID, entity_type: str) -> Optional[Entity]:
```
Comment (sebhmg) [inline geoh5py/workspace/workspace.py:741]: 👍

File: tests/remove_root_test.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+        workspace.finalize()
+
+        # Remove the root
+        project = File(h5file_path, "r+")
+        base = list(project.keys())[0]
+        del project[base]["Root"]
+        del project[base]["Groups"]
+        del project[base]["Types"]["Group types"]
+        project.close()
```
Comment (sebhmg) [inline tests/remove_root_test.py:65]: **suggestion (RAII):** use context `with ...`

**Review by sebhmg — CHANGES_REQUESTED** [review]
looks good overall ! With some nice refactoring. Thanks.

Please see suggestions and questions.
Also, I have pointed to a couple of changes that seem unrelated and should maybe not belong to this PR.

File: geoh5py/io/h5_reader.py [hunk 6]
```diff
@@ -276,6 +276,30 @@ def fetch_property_groups(
...
+        Fetch all uuids of a given type from geoh5
+
+        :param h5file: Name of the target geoh5 file
+        :param entity_type: Type of entity from
+            'group', 'data', 'object', 'group_type', 'data_type', 'object_type'
+
+        :return uuids: [uuid1, uuid2, ...]
+            List of uuids
+        """
+        project = h5py.File(h5file, "r")
```
[the remark this answers is not in the document]
Comment (domfournier) [reply geoh5py/io/h5_reader.py:291]: Yes, will change throughout

File: geoh5py/io/h5_reader.py [hunk 7]
```diff
@@ -276,6 +276,30 @@ def fetch_property_groups(
...
+        name = list(project.keys())[0]
+
+        entity_type = entity_type.capitalize()
+        if entity_type in ["Group", "Object"]:
+            entity_type += "s"
```
[the remark this answers is not in the document]
Comment (domfournier) [reply geoh5py/io/h5_reader.py]: Done

File: geoh5py/objects/octree.py (hunk 1 above)
Comment (domfournier) [reply geoh5py/objects/octree.py:199]: oh please, I have been so good recently. Please let that one slide, please?

File: poetry.lock (hunk 2 above)
Comment (domfournier) [reply poetry.lock:9]: Yea not sure again, git extension keeps bugging me again.

File: geoh5py/workspace/workspace.py (hunk 3 above)
Comment (domfournier) [reply geoh5py/workspace/workspace.py:107]: Done

File: geoh5py/objects/octree.py (hunk 1 above)
Comment (sebhmg) [reply geoh5py/objects/octree.py:199]: alright

File: tests/remove_root_test.py (hunk 5 above)
Comment (domfournier) [reply tests/remove_root_test.py:65]: Done

File: geoh5py/shared/utils.py [hunk 8]
```diff
@@ -16,11 +16,28 @@
...
+    Open in read+ mode a geoh5 file from string.
+
+    :param file: Name or handle to a geoh5 file.
+
+    :return h5py.File: Handle to an opened h5py file.
+    """
+    if isinstance(file, h5py.File):
+        return file
+
+    return h5py.File(file, "r+")
```
Comment (sebhmg) [inline geoh5py/shared/utils.py:38]: see https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager to implement context and be able to use: `with fetch_h5_handle(file) as h5file:`
with no need to worry about closing the file or not

```suggestion
@contextmanager
def fetch_h5_handle(
    file: Union[str, h5py.File],
) -> h5py.File:
    """
    Open in read+ mode a geoh5 file from string.
    If receiving a file instead of a string, merely return the given file.

    :param file: Name or handle to a geoh5 file.

    :return h5py.File: Handle to an opened h5py file.
    """
    if isinstance(file, h5py.File):
        try:
            return file
        finally:
            pass
    else:
        h5file = h5py.File(file, "r+")
        try:
            yield h5file
        finally:
            h5file.close()
```

File: geoh5py/workspace/workspace.py [hunk 9]
```diff
@@ -130,17 +127,17 @@ def active() -> Workspace:
         # so that type check does not complain of possible returned None
         return cast(Workspace, active_one)
 
-    def _all_data(self) -> List["data.Data"]:
+    def _all_data(self) -> List["Entity"]:
```
Comment (sebhmg) [inline geoh5py/workspace/workspace.py:130]: why using the more generic type `Entity` when it is returning `Data`?

File: geoh5py/workspace/workspace.py [hunk 10]
```diff
@@ -130,17 +127,17 @@ def active() -> Workspace:
...
         """Get all active Data entities registered in the workspace."""
         self.remove_none_referents(self._data, "Data")
         return [cast("data.Data", v()) for v in self._data.values()]
 
-    def _all_groups(self) -> List["group.Group"]:
+    def _all_groups(self) -> List["Entity"]:
```
Comment (sebhmg) [inline geoh5py/workspace/workspace.py:135]: why using the more generic type `Entity` when it is returning `Group`?

File: geoh5py/workspace/workspace.py [hunk 11]
```diff
@@ -130,17 +127,17 @@ def active() -> Workspace:
...
         """Get all active Group entities registered in the workspace."""
         self.remove_none_referents(self._groups, "Groups")
         return [cast("group.Group", v()) for v in self._groups.values()]
 
-    def _all_objects(self) -> List["object_base.ObjectBase"]:
+    def _all_objects(self) -> List["Entity"]:
```
Comment (sebhmg) [inline geoh5py/workspace/workspace.py:140]: why using the more generic type `Entity` when it is returning Objects?

File: geoh5py/workspace/workspace.py [hunk 12]
```diff
@@ -378,42 +375,48 @@ def create_object_or_group(
         return None
 
     @property
-    def data(self) -> List["data.Data"]:
+    def data(self) -> List["Entity"]:
```
Comment (sebhmg) [inline geoh5py/workspace/workspace.py:378]: why using the more generic type `Entity` when it is returning `Data`?

File: geoh5py/workspace/workspace.py [hunk 13]
```diff
@@ -690,7 +722,7 @@ def get_entity(self, name: Union[str, uuid.UUID]) -> List[Optional[Entity]]:
         return entity_list
 
     @property
-    def groups(self) -> List["group.Group"]:
+    def groups(self) -> List["Entity"]:
```
Comment (sebhmg) [inline geoh5py/workspace/workspace.py:725]: why using the more generic type `Entity` when it is returning `Group`?

File: geoh5py/workspace/workspace.py [hunk 14]
```diff
@@ -854,6 +900,18 @@ def workspace(self):
         """
         return self
 
+    def _io_call(self, file, fun, *args, **kwargs):
+        """
+        Run a H5Writer or H5Reader function with validation of target geoh5
+        """
+        h5file = self.validate_h5file(file)
```
Comment (sebhmg) [inline geoh5py/workspace/workspace.py:907]: looks like `validate_h5file` should also be decorated with `@contextmanager`

File: geoh5py/workspace/workspace.py [hunk 15]
```diff
@@ -787,7 +830,7 @@ def name(self) -> str:
         return self._name
 
     @property
-    def objects(self) -> List["object_base.ObjectBase"]:
+    def objects(self) -> List["Entity"]:
```
Comment (sebhmg) [inline geoh5py/workspace/workspace.py:833]: why using the more generic type `Entity` when it is returning Objects?

File: geoh5py/workspace/workspace.py [hunk 16]
```diff
@@ -810,32 +853,35 @@ def root(self) -> Optional[Entity]:
...
 
+    def validate_h5file(self, file) -> h5py.File:
+        """
+        Validate the h5file name
+        """
+        if file is None:
+            file = self.h5file
+
+        return fetch_h5_handle(file)
+
```
Comment (sebhmg) [inline geoh5py/workspace/workspace.py:884]: this function is not necessary and prevents usage of context manager.
Can be replaced everywhere by `fetch_h5_handle(file or self.h5file)`

With `@contextmanager` on `fetch_h5_handle`, it becomes everywhere:
`with fetch_h5_handle(file or self.h5file) as h5file:`

**Review by sebhmg — COMMENTED** [review]
some issue around usage of `fetch_h5_handle`

Please see comments and how to take advantage of context manager, to retin the conveniance `fetch_h5_handle` withou hacing to test whterh to close or not

**Review by sebhmg — CHANGES_REQUESTED** [review]
some issue around usage of `fetch_h5_handle`

Please see comments and how to take advantage of context manager, to retain the convenience of `fetch_h5_handle` without having to test whether to close or not

Also questions about modified returned types.

File: geoh5py/workspace/workspace.py (hunk 12 above)
Comment (domfournier) [reply geoh5py/workspace/workspace.py:378]: Same for all the comments below, I am now getting an error from mypy regarding the mix of types
image

Any idea?

Comment (domfournier) [reply geoh5py/workspace/workspace.py:378]: found this https://stackoverflow.com/questions/56738485/why-do-i-get-a-warning-when-concatenating-lists-of-mixed-types-in-pycharm

Looks like I would need to avoid adding the lists....

Comment (sebhmg) [reply geoh5py/workspace/workspace.py:378]: I would think you just need to cast when adding them:
`my_base_type_list = cast(List[base_type],my_derived_type1_list) + cast(List[base_type],my_derived_type2_list)`

Comment (domfournier) [reply geoh5py/workspace/workspace.py:378]: Done

File: geoh5py/io/h5_reader.py [hunk 17]
```diff
@@ -60,183 +64,226 @@ def fetch_attributes(
...
-                type_attributes["entity_type"]["color_map"]["values"] = entity["Type"][
-                    "Color map"
-                ][:]
-
-            if "Value map" in entity["Type"].keys():
-                value_map = entity["Type"]["Value map"][:]
-                mapping = {}
-                for key, value in value_map.tolist():
-                    value = cls.str_from_utf8_bytes(value)
+        h5file = fetch_h5_handle(file)
```
[the remark this answers is not in the document]
Comment (domfournier) [reply geoh5py/io/h5_reader.py]: Love it. Done

File: geoh5py/workspace/workspace.py [hunk 18]
```diff
@@ -873,14 +873,14 @@ def types(self) -> List["EntityType"]:
...
-    def validate_h5file(self, file) -> h5py.File:
+    def validate_file(self, file) -> h5py.File:
         """
         Validate the h5file name
         """
         if file is None:
             file = self.h5file
 
-        return fetch_h5_handle(file)
+        return file
```
Comment (sebhmg) [inline geoh5py/workspace/workspace.py:883]: 👍

**Review by sebhmg — APPROVED** [review]
launch
